tests/SchedulerTest.php

```diff
@@ -0,0 +1,221 @@
+            })
+            ->on(CountableScheduler::ON_TASK_FAILED, function (Task $t, Throwable $_) use (&$taskFailed) {
+                $taskFailed = true;
+            });
+
+        $this->runOff();
+
+        $this->assertTrue($taskFailed, 'Scheduler could not handle failed tasks');
+    }
+
+    public function testTaskRunSuccess()
+    {
+        $taskFulFilled = false;
+        $deferred = new Promise\Deferred();
+        $task = new PromiseBoundTask($deferred->promise());
+        $this->scheduler
+            ->schedule($task, new ImmediateDueFrequency())
+            ->on(CountableScheduler::ON_TASK_RUN, function (Task $_, ExtendedPromiseInterface $p) use ($deferred) {
+                $deferred->resolve(10);
+            })
+            ->on(CountableScheduler::ON_TASK_DONE, function (Task $_, $any) use (&$taskFulFilled) {
+                $taskFulFilled = true;
+            });
+
+        $this->runOff();
+
+        $this->assertTrue($taskFulFilled, 'Scheduler could not run task successfully');
+    }
+
+    public function testPropagateTaskExceptions()
+    {
+        $this->expectException(TaskRejectedException::class);
+
+        $exception = null;
+        $deferred = new Promise\Deferred();
+        $task = new PromiseBoundTask($deferred->promise());
+
+        $this->scheduler
+            ->schedule($task, new ImmediateDueFrequency())
+            ->on(CountableScheduler::ON_TASK_RUN, function (Task $t, ExtendedPromiseInterface $_) use ($deferred) {
+                $deferred->reject(new TaskRejectedException('Rejected!'));
+            })
+            ->on(CountableScheduler::ON_TASK_FAILED, function (Task $_, Throwable $reason) use (&$exception) {
+                $exception = $reason;
+            });
+
+        $this->runOff();
+
+        throw $exception;
+    }
+
+    public function testSpawningNewTasks()
+    {
+        $task1 = new PromiseBoundTask(Promise\resolve());
+        $task2 = new PromiseBoundTask(Promise\resolve());
+
+        $this->scheduler
+            ->schedule($task1, new ImmediateDueFrequency())
+            ->schedule($task2, new ImmediateDueFrequency());
+
+        $this->runOff();
+
+        $this->assertEquals(
+            2,
+            $this->scheduler->count(),
+            'Scheduler::schedule() failed to schedule expected tasks'
+        );
+    }
+
+    public function testCountScheduledTimers()
+    {
+        $task1 = new PromiseBoundTask(Promise\resolve());
+        $task2 = new PromiseBoundTask(Promise\resolve());
+
+        $this->scheduler
+            ->schedule($task1, new ImmediateDueFrequency())
+            ->schedule($task2, new ImmediateDueFrequency());
+
+        $this->runOff();
+
+        $this->assertEquals(
+            2,
+            $this->scheduler->countTimers(),
+            'Scheduler::schedule() could not return expected scheduled timers'
+        );
+    }
+
+    public function testTimersAndTasksCountAfterCancellation()
```

Comment:
Instead of having a separate method for a specific use case I would test every count in all the other tests.